### DISCUSSION TITLE
Any way to run the plugin without Maven or as a Standalone JAR?

### DISCUSSION THREAD

**Discussion by darshitpp** [revelc/formatter-maven-plugin#696] [Q&A]
My project(s), unfortunately, doesn't use Maven, but I'd like the flexibility that this plugin provides. Any way to run it as a standalone app which I can trigger from a Git pre-commit hook? Thanks!

**Comment by hazendaz** [comment]
It's feasible to use maven to run the plugin in git commit hook.  Haven't done so myself but is 9n my plan to do so at some point.  Think you would just run the format cycle giving full gav if maven 3.8.7 or short for maven 3.9.0.  Typically short like mvn formatter:format.  If I get something before you do I'll document it here.  In theory maven wrapper likely would be a better fit so you don't need to install it.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Darshit Patel ***@***.***>
Sent: Tuesday, February 28, 2023 5:56:37 AM
To: revelc/formatter-maven-plugin ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [revelc/formatter-maven-plugin] Any way to run the plugin without Maven or as a Standalone JAR? (Discussion #696)


My project(s), unfortunately, doesn't use Maven, but I'd like the flexibility that this plugin provides. Any way to run it as a standalone app which I can trigger from a Git pre-commit hook? Thanks!

**Reply by darshitpp** [reply]
This doesn't work because it requires my project to be a maven project with a POM to even run the goal.

```
[INFO] ------------------< org.apache.maven:standalone-pom >-------------------
[INFO] Building Maven Stub Project (No POM) 1
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.114 s
[INFO] Finished at: 2023-03-03T13:33:48+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal net.revelc.code.formatter:formatter-maven-plugin:2.16.0:format (default-cli): Goal requires a project to execute but there is no POM in this directory (<path_to_project>). Please verify you invoked Maven from the correct directory. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MissingProjectException
```

**Comment by hazendaz** [comment]
Please try again with latest and show stack.  We are a small team and don't really support anything but latest.  Once confirmed can look at making it not require maven project.  Thanks.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by darshitpp** [comment]
We don't really have JDK 11 installed to run with 2.17.x versions of the libraries. I am using the latest 2.16.0 JAR. Not much of a stack trace to share, unfortunately. I use the install-file goal to load the plugin into the the local maven repository, and then simply run the command. Here's the trace:
```bash
mvn org.apache.maven.plugins:maven-install-plugin:2.5.2:install-file -Dfile=.git/hooks/formatter-maven-plugin-2.16.0.jar
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------< org.apache.maven:standalone-pom >-------------------
[INFO] Building Maven Stub Project (No POM) 1
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- maven-install-plugin:2.5.2:install-file (default-cli) @ standalone-pom ---
[INFO] Installing /path_to_plugin/formatter-maven-plugin-2.16.0.jar to /path_to_m2/.m2/repository/net/revelc/code/formatter/formatter-maven-plugin/2.16.0/formatter-maven-plugin-2.16.0.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.814 s
[INFO] Finished at: 2023-03-08T10:46:25+05:30
[INFO] ------------------------------------------------------------------------

mvn net.revelc.code.formatter:formatter-maven-plugin:2.16.0:format -DconfigFile=/path_to_formatter/format.xml -DsourceDirectory=/path_to_source/src/ -Dincludes=/path_to_single_file_test/Test.java
+ mvn net.revelc.code.formatter:formatter-maven-plugin:2.16.0:format -DconfigFile=/path_to_formatter/format.xml -DsourceDirectory=/path_to_source/src/ -Dincludes=/path_to_single_file_test/Test.java
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------< org.apache.maven:standalone-pom >-------------------
[INFO] Building Maven Stub Project (No POM) 1
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.117 s
[INFO] Finished at: 2023-03-08T10:46:26+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal net.revelc.code.formatter:formatter-maven-plugin:2.16.0:format (default-cli): Goal requires a project to execute but there is no POM in this directory (/current_directory_which_has_projects_but_no_POM). Please verify you invoked Maven from the correct directory. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MissingProjectException
```

The above is generated when the git hook is run. I have hardcoded the values of the `configFile`, `sourceDirectory`, and `includes` for testing and verification. 

I'd be happy to contribute if you require help in making the plugin non-maven based. Would be great to learn, and contribute to something new!

**Reply by ctubbsii** [reply]
> I'd be happy to contribute if you require help in making the plugin non-maven based. Would be great to learn, and contribute to something new!

The entire point of the plugin is to wrap the Eclipse formatter for Maven projects. If you need to run outside of Maven, you could use the Eclipse formatter directly. It doesn't make a lot of sense to try to extend this project to work for non-Maven use cases, since the whole point of this plugin's existence is to be a wrapper specifically to serve Maven projects. So, it will always be Maven-based, at least in the sense that it requires Maven to execute (even if the target folder it is formatting isn't a Maven project).

As for executing outside of a Maven project, to format a directory, there was a lengthy discussion on #261 for why making it only work within a Maven project was done. It was needed in order to support multi-module Maven projects better. In that discussion, it was suggested a separate "format-directory" or similar new goal could be created to run the formatting outside of a Maven project. That's still an option, and one possible way you can contribute. But, we would only add that feature to new versions... which require JDK 11 at least. So I'm not sure such a change would directly help you if you aren't going to have that installed.

**Reply by darshitpp** [reply]
Got it. 

Let me try it anyway, I can use JDK 11 on my personal dev system. Perhaps I can learn a new thing or two

**Reply by darshitpp** [reply]
Any docs that I can refer to regarding building/debugging Maven plugins? Passing `-DsourceDirectory` during the goal execution does get the correct value from `System.getProperty` but somehow doesn't reflect in the `private File sourceDirectory;` variable.

**Reply by ctubbsii** [reply]
I don't know. The Maven project might have some docs. Usually I learn by looking at other projects/existing code.

**Reply by darshitpp** [reply]
I understand that. But the behaviour on my system is weird, and inexplicable. Google searching could only help so much. Here's where I'm stuck at

1. Created a new Mojo named `DirectoryFormatterMojo`
2. The new Mojo has exactly the same things as the FormatterMojo, except that I kept the default values as in `sourceDirectory`, `testSourceDirectory`, and `directories` as `.` (current directory)
3. Passing the value through CLI as `-DsourceDirectory` does not inject the value in the `sourceDirectory` variable.
4. I am able to inject the value of the path in the `directories` by specifying the path as `../../../../../<path>`
5. Doing the same with `sourceDirectory` does not have the same behaviour as with the `directories` above.

This has left me quite confused.

**Reply by ctubbsii** [reply]
Without seeing your code, I can't really help. You can run things with -X to get a full output of debugging log messages, etc., and to see what is being passed as configuration to the plugin.

**Comment by delanym** [comment]
@darshitpp before this plugin I used to use a shell function that invokes eclipse directly

``` sh
eformat(){
  eclipse -nosplash -data /home/sol/eclipse-workspace -vm java -application org.eclipse.jdt.core.JavaCodeFormatter -verbose -config /git/org.eclipse.jdt.core.prefs $@
}
```

Then I just run `eformat <directory>` and it formats everything recursively.
